src/nupic/ntypes/Sdr.hpp

```diff
@@ -0,0 +1,273 @@
+
+using namespace std;
+
+namespace nupic {
+
+/**
+ * SparseDistributedRepresentation in C++
+ *
+ * ### Description
+ * This class manages the specification and momentary value of a Sparse
+ * Distributed Representation (SDR).  An SDR is a group of boolean values which
+ * represent the state of a group of neurons or their associated processes. 
+ *
+ * This class automatically converts between the commonly used SDR data formats:
+ * which are dense, index, and flat-index.  Converted values are cached by this
+ * class, so getting a value in one format many times incurs no extra
+ * performance cost.  Assigning to the SDR via a setter method will clear these
+ * cached values and cause them to be recomputed as needed.
+ *
+ *    Dense Format: A contiguous array of boolean values, representing all of
+ *    the bits in the SDR.  This format allows random-access queries of the SDRs
+ *    values.
+ *
+ *    Index Format: A list of lists of the indices of the true bits in the SDR.
+ *    The outter list contains an entry for each dimension in the SDR.  The
+ *    inner lists contain the coordinates of each true bit.  The inner lists run
+ *    in parallel. This format is useful because it contains the location of
+ *    each true bit inside of the SDR's dimensional space.
+ *
+ *    Flat Index Format: A list of the indices of the true bits, indexed into
+ *    the the flattened SDR.  This format allows for quickly accessing all of
+ *    the true bits in the SDR.
+ *
+ * Flat Formats: This class uses C-order throughout, meaning that when iterating
+ * through the SDR, the last/right-most index changes fastest.
+ *
+ * Boolean Data Type: Although the documentation for this class may refer to the
+ * values in an SDR as bits they are in fact stored as bytes (of type signed
+ * character).
+ *
+ * Example usage:
+ *
+ *    // Make an SDR with 10,000 bits, arranged in a (100 x 100) gird.
+ *    SparseDistributedRepresentation X({100, 100});
+ *
+ *    // Assign data in any format.
+ *    X.setFlatIndex( mySparseData ); // Assigns mySparseData to SDR's value.
+ *    X.setDense( myDenseData );      // myDenseData overwrites mySparseData.
+ *
+ *    // Access data in different formats.
+ *    indices      = X.getFlatIndex(); // Calculated from myDenseData.
+ *    indicesAgain = X.getFlatIndex(); // Reuses previous result, no work done.
+ *
+ */
+class SparseDistributedRepresentation : public Serializable
+{
+public:
+    /**
+     * Create an uninitialized SDR.  The only way to initialize SDRs created by
+     * this method is by calling SDR.load().  Until SDR.load() is called this
+     * SDR is in an unusable state.
+     */
+    SparseDistributedRepresentation();
+
+    /**
+     * Create an SDR object.  Initially this SDR has no value set.
+     *
+     * @param dimensions A list of dimension sizes, defining the shape of the
+     * SDR.
+     */
+    SparseDistributedRepresentation(const vector<UInt> dimensions);
```

Comment:
This class does not force anyone to use multiple dimensions, see dense & flatIndex.

I think more code would use dimensions if the tools for working with dimensions were easier to use, which is one of the goals for this API.  Perhaps I'm just too accustomed to using Numpy, which BTW is my golden standard of dimensional tools.

Reply:
`This class does not force anyone ...`

Understood.  My point is that an algorithm can view a block of data with any dimensions it wants to perhaps viewing it multiple ways at the same time but the underlining space is still flat.  I guess what I object to is the current implementation of Link where dimensions are required.  

I envision this class to be sort of like numpy where internally its just a block of space but externally algorithms can use the facilities of numpy to iterate through the space in different ways or dimensions without requiring any modifications to the structure underneath.  The C++ algorithms don't do that currently.

Reply:
> `to iterate through the space in different ways or dimensions without requiring any modifications to the structure underneath`

The SDR class can help with that, albeit with an extra data copy.  Simply make a new SDR with the dimensions you want, and assign the flat-index or the dense data across.  Then access the data in the index format which converts from 1D to the correct dimensions.  Neither flatIndex nor dense data formats care about the actual dimensions as long as the SDR's total size matches.
```
SDR a({100, 100, 3});
a.setDense( /* insert data here */ );

SDR b({10000, 3});    // Merge the first 2 dimensions.
b.setFlatIndex( a.getFlatIndex() );
auto index = b.getIndex();
for(UInt idx = 0; idx < index->at(0).size(); idx++) {
    for(UInt dim = 0; dim < 2; dim++) {
        auto coordinate = index[dim][idx];
   }
}
```
This class could (but does not currently) allow you to query a location using an index-tuple.  That could look like:
```
SDR a({3, 3});
a.setDense( /* insert data here */ );
a.at( {1, 2} );
// This would be equivalent to a.getDense()->at(5);
```

Reply:
I like this idea of making n-dimensional SDRs more accessible. This example would allow is to reduce `CoordinateConverter_{1D, 2D, ND}` from SP class